Race condition in setBuf

```
==================
WARNING: DATA RACE
Read at 0x00c019349648 by goroutine 35:
  github.com/dgraph-io/ristretto.(*Cache).Set()
      /home/pawan0201/go/pkg/mod/github.com/dgraph-io/ristretto@v0.0.2-0.20200115201040-8f368f2f2ab3/cache.go:205 +0x1ed
  github.com/dgraph-io/badger/v2/table.(*Table).block()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/table/table.go:459 +0xe9a
  github.com/dgraph-io/badger/v2/table.(*Iterator).next()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/table/iterator.go:296 +0x216
  github.com/dgraph-io/badger/v2/table.(*Iterator).next()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/table/iterator.go:311 +0x699
  github.com/dgraph-io/badger/v2/table.(*Iterator).Next()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/table/iterator.go:367 +0x70
  github.com/dgraph-io/badger/v2/table.(*ConcatIterator).Next()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/table/iterator.go:485 +0x59
  github.com/dgraph-io/badger/v2/table.(*node).next()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/table/merge_iterator.go:82 +0xce
  github.com/dgraph-io/badger/v2/table.(*MergeIterator).Next()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/table/merge_iterator.go:158 +0x6a
  github.com/dgraph-io/badger/v2/table.(*node).next()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/table/merge_iterator.go:80 +0x73
  github.com/dgraph-io/badger/v2/table.(*MergeIterator).Next()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/table/merge_iterator.go:158 +0x6a
  github.com/dgraph-io/badger/v2.(*Iterator).parseItem()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/iterator.go:648 +0x933
  github.com/dgraph-io/badger/v2.(*Iterator).Next()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/iterator.go:561 +0x1fd
  github.com/dgraph-io/badger/v2.TestDropAllWithPendingTxn.func3()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/managed_db_test.go:218 +0x183

Previous write at 0x00c019349648 by goroutine 82:
  github.com/dgraph-io/ristretto.(*Cache).Clear()
      /home/pawan0201/go/pkg/mod/github.com/dgraph-io/ristretto@v0.0.2-0.20200115201040-8f368f2f2ab3/cache.go:254 +0x9f
  github.com/dgraph-io/badger/v2.(*DB).dropAll()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/db.go:1511 +0x57e
  github.com/dgraph-io/badger/v2.(*DB).DropAll()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/db.go:1467 +0x47
  github.com/dgraph-io/badger/v2.TestDropAllWithPendingTxn.func4()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/managed_db_test.go:248 +0x7f

Goroutine 35 (running) created at:
  github.com/dgraph-io/badger/v2.TestDropAllWithPendingTxn()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/managed_db_test.go:210 +0x3e6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 82 (running) created at:
  github.com/dgraph-io/badger/v2.TestDropAllWithPendingTxn()
      /home/pawan0201/go/src/github.com/dgraph-io/badger/managed_db_test.go:245 +0x43f
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199
==================
badger 2020/02/05 10:11:51 INFO: Resuming writes
badger 2020/02/05 10:11:51 INFO: Got compaction priority: {level:0 score:1.73 dropPrefix:[]}
--- FAIL: TestDropAllWithPendingTxn (8.77s)
    managed_db_test.go:228: Got number of keys: 10000
    managed_db_test.go:224: Got error during value copy: Invalid value pointer offset: 3689714 greater than current offset: 20
    testing.go:853: race detected during execution of test
```

The line in question is 
https://github.com/dgraph-io/ristretto/blob/8f368f2f2ab3a54cbe62fb9772cd75ce55e07802/cache.go#L205 and 

https://github.com/dgraph-io/ristretto/blob/8f368f2f2ab3a54cbe62fb9772cd75ce55e07802/cache.go#L254
https://teamcity.dgraph.io/viewLog.html?buildId=43004&buildTypeId=Badger_UnitTests revealed the race condition

--- a/cache.go
+++ b/cache.go
@@ -23,7 +23,6 @@ import (
 	"bytes"
 	"errors"
 	"fmt"
-	"sync"
 	"sync/atomic"
 	"time"
 
@@ -51,9 +50,6 @@ type Cache struct {
 	// setBuf is a buffer allowing us to batch/drop Sets during times of high
 	// contention.
 	setBuf chan *item
-	// setBufLock is a reader-writer lock to ensure there are no race conditions
-	// when calling one of Set/Del and one of Clear/Close simultaneously.
-	setBufLock sync.RWMutex
 	// onEvict is called for item evictions.
 	onEvict onEvictFunc
 	// KeyToHash function is used to customize the key hashing algorithm.
@@ -234,8 +230,6 @@ func (c *Cache) SetWithTTL(key, value interface{}, cost int64, ttl time.Duration
 		i.flag = itemUpdate
 	}
 	// Attempt to send item to policy.
-	c.setBufLock.RLock()
-	defer c.setBufLock.RUnlock()
 	select {
 	case c.setBuf <- i:
 		return true
@@ -257,13 +251,11 @@ func (c *Cache) Del(key interface{}) {
 	// So we must push the same item to `setBuf` with the deletion flag.
 	// This ensures that if a set is followed by a delete, it will be
 	// applied in the correct order.
-	c.setBufLock.RLock()
 	c.setBuf <- &item{
 		flag:     itemDelete,
 		key:      keyHash,
 		conflict: conflictHash,
 	}
-	c.setBufLock.RUnlock()
 }
 
 // Close stops all goroutines and closes all channels.
@@ -275,12 +267,7 @@ func (c *Cache) Close() {
 	c.stop <- struct{}{}
 	close(c.stop)
 	c.stop = nil
-
-	// Close setBuf channel.
-	c.setBufLock.Lock()
 	close(c.setBuf)
-	c.setBufLock.Unlock()
-
 	c.policy.Close()
 }
 
@@ -294,10 +281,15 @@ func (c *Cache) Clear() {
 	// Block until processItems goroutine is returned.
 	c.stop <- struct{}{}
 
-	// Swap out the setBuf channel.
-	c.setBufLock.Lock()
-	c.setBuf = make(chan *item, setBufSize)
-	c.setBufLock.Unlock()
+	// Clear out the setBuf channel.
+loop:
+	for {
+		select {
+		case <-c.setBuf:
+		default:
+			break loop
+		}
+	}
 
 	// Clear value hashmap and policy data.
 	c.policy.Clear()
@@ -315,10 +307,6 @@ func (c *Cache) processItems() {
 	for {
 		select {
 		case i := <-c.setBuf:
-			// There is no need to protect setBuf with the RWMutex here because both Close
-			// and Clear make sure processItems is not running before writing or closing
-			// the channel.
-
 			// Calculate item cost value if new or update.
 			if i.cost == 0 && c.cost != nil && i.flag != itemDelete {
 				i.cost = c.cost(i.value)

--- a/cache.go
+++ b/cache.go
@@ -23,6 +23,7 @@ import (
 	"bytes"
 	"errors"
 	"fmt"
+	"sync"
 	"sync/atomic"
 	"time"
 
@@ -50,6 +51,9 @@ type Cache struct {
 	// setBuf is a buffer allowing us to batch/drop Sets during times of high
 	// contention.
 	setBuf chan *item
+	// setBufLock is a reader-writer lock to ensure there are no race conditions
+	// when calling one of Set/Del and one of Clear/Close simultaneously.
+	setBufLock sync.RWMutex
 	// onEvict is called for item evictions.
 	onEvict onEvictFunc
 	// KeyToHash function is used to customize the key hashing algorithm.
@@ -230,6 +234,8 @@ func (c *Cache) SetWithTTL(key, value interface{}, cost int64, ttl time.Duration
 		i.flag = itemUpdate
 	}
 	// Attempt to send item to policy.
+	c.setBufLock.RLock()
+	defer c.setBufLock.RUnlock()
 	select {
 	case c.setBuf <- i:
 		return true
@@ -251,11 +257,13 @@ func (c *Cache) Del(key interface{}) {
 	// So we must push the same item to `setBuf` with the deletion flag.
 	// This ensures that if a set is followed by a delete, it will be
 	// applied in the correct order.
+	c.setBufLock.RLock()
 	c.setBuf <- &item{
 		flag:     itemDelete,
 		key:      keyHash,
 		conflict: conflictHash,
 	}
+	c.setBufLock.RUnlock()
 }
 
 // Close stops all goroutines and closes all channels.
@@ -267,7 +275,12 @@ func (c *Cache) Close() {
 	c.stop <- struct{}{}
 	close(c.stop)
 	c.stop = nil
+
+	// Close setBuf channel.
+	c.setBufLock.Lock()
 	close(c.setBuf)
+	c.setBufLock.Unlock()
+
 	c.policy.Close()
 }
 
@@ -280,8 +293,12 @@ func (c *Cache) Clear() {
 	}
 	// Block until processItems goroutine is returned.
 	c.stop <- struct{}{}
+
 	// Swap out the setBuf channel.
+	c.setBufLock.Lock()
 	c.setBuf = make(chan *item, setBufSize)
+	c.setBufLock.Unlock()
+
 	// Clear value hashmap and policy data.
 	c.policy.Clear()
 	c.store.Clear()
@@ -298,6 +315,10 @@ func (c *Cache) processItems() {
 	for {
 		select {
 		case i := <-c.setBuf:
+			// There is no need to protect setBuf with the RWMutex here because both Close
+			// and Clear make sure processItems is not running before writing or closing
+			// the channel.
+
 			// Calculate item cost value if new or update.
 			if i.cost == 0 && c.cost != nil && i.flag != itemDelete {
 				i.cost = c.cost(i.value)

